src/gpio.rs

```diff
@@ -372,123 +353,195 @@ impl<Gpio, Index, Mode> Pin<Gpio, Index, Mode> {
 
 impl<Gpio, Index, Mode> Pin<Gpio, Index, Mode>
 where
-    Gpio: marker::GpioStatic,
+    Gpio: marker::Gpio,
     Index: marker::Index,
 {
     /// Configures the pin to operate as an input pin
-    pub fn into_input(self, moder: &mut Gpio::MODER) -> Pin<Gpio, Index, Input> {
-        moder.input(self.index.index());
+    pub fn into_input(self) -> Pin<Gpio, Index, Input> {
+        // NOTE(unsafe) atomic modify with no side effects
+        unsafe { (*self.gpio.ptr()).input(self.index.index()) };
         self.into_mode()
     }
 
     /// Convenience method to configure the pin to operate as an input pin
     /// and set the internal resistor floating
-    pub fn into_floating_input(
-        self,
-        moder: &mut Gpio::MODER,
-        pupdr: &mut Gpio::PUPDR,
-    ) -> Pin<Gpio, Index, Input> {
-        moder.input(self.index.index());
-        pupdr.floating(self.index.index());
+    pub fn into_floating_input(self) -> Pin<Gpio, Index, Input> {
+        // NOTE(unsafe) atomic modify with no side effects
+        unsafe {
+            (*self.gpio.ptr()).floating(self.index.index());
+            (*self.gpio.ptr()).input(self.index.index());
+        }
         self.into_mode()
     }
 
     /// Convenience method to configure the pin to operate as an input pin
     /// and set the internal resistor pull-up
-    pub fn into_pull_up_input(
-        self,
-        moder: &mut Gpio::MODER,
-        pupdr: &mut Gpio::PUPDR,
-    ) -> Pin<Gpio, Index, Input> {
-        moder.input(self.index.index());
-        pupdr.pull_up(self.index.index());
+    pub fn into_pull_up_input(self) -> Pin<Gpio, Index, Input> {
+        // NOTE(unsafe) atomic modify with no side effects
```

Comment:
I'm not sure if this comment is a little misleading or could be misunderstood. 

First, the thing, which makes it unsafe is not the modification (which it sounds like in the comment) but rather the dereferencing of the pointer. Secondly, to add to the confusion, the method (e.g. `pull_up`) is not marked as unsafe.

I'd propose marking the functions from `gpio_reg_trait!` as `unsafe` and additionally moving the `ptr()` there as well? (if this is technically possible). 

Then document under which circumstances these functions can be used safely ("Is only safe to call, when mutable access to peripheral is guaranteed" or something similar, which makes sense in that context) and **here** you would then write why the circumstances, which the unsafe function defines to be safe usage, do indeed exist.